README.md
# ExAws.STS

[![Module Version](https://img.shields.io/hexpm/v/ex_aws_sts.svg)](https://hex.pm/packages/ex_aws_sts)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/ex_aws_sts/)
[![Total Download](https://img.shields.io/hexpm/dt/ex_aws_sts.svg)](https://hex.pm/packages/ex_aws_sts)
[![License](https://img.shields.io/hexpm/l/ex_aws_sts.svg)](https://github.com/ex-aws/ex_aws_sts/blob/master/LICENSE)
[![Last Updated](https://img.shields.io/github/last-commit/ex-aws/ex_aws_sts.svg)](https://github.com/ex-aws/ex_aws_sts/commits/master)

Service module for [https://github.com/ex-aws/ex_aws](https://github.com/ex-aws/ex_aws).

## Installation

The package can be installed by adding `:ex_aws_sts` to your list of dependencies in `mix.exs` along with `:ex_aws`.

```elixir
def deps do
  [
    {:ex_aws, "~> 2.2"},
    {:ex_aws_sts, "~> 2.2"},
  ]
end
```

Documentation can be found at [https://hexdocs.pm/ex_aws_sts](https://hexdocs.pm/ex_aws_sts).

## Role based authentication

Using the `:awscli_auth_adapter` option of `ex_aws` is supported, but requires additional dependencies:

```elixir
{:sweet_xml, "~> 0.6"}
{:configparser_ex, "~> 2.0"}
```

### Using AWS CLI config file with source profile

`ExAws.STS` allows to authentication based on `role_arn` and `source_profile` as specified in the `awscli` config file.

When specified in your `~/.aws/config` you can set 
 
```elixir
config :ex_aws,
  secret_access_key: [{:awscli, "profile_name", 1_800}],
  access_key_id: [{:awscli, "profile_name", 1_800}],
  awscli_auth_adapter: ExAws.STS.AuthCache.AssumeRoleCredentialsAdapter
```

and if the profile `profile_name` sets a `role_arn` then this will make ExAws
issue an `AssumeRoleCredentials` request to fetch the `access_key_id`
and `secret_access_key`.
The third element in the tuple (1_800 in the example) is the desired expiration in seconds.
[AWS enforces a minimum](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html#API_AssumeRole_RequestParameters) of 900 seconds (15 minutes).

### Using AWS CLI config from ENV vars

It is possible to inject the credentials by configuration, for example, by using ENV vars. This is very useful for containerized applications.

In order to do that, just place in the config the block `awscli_credentials` with your `profile_name` as key and the corresponding values. Then under `access_key_id` and `secret_access_key` just make a reference to the profile, so it will be used to ask for the credentials automatically.

```elixir
config :ex_aws,
  access_key_id: [{:awscli, "default", 1_800}],
  secret_access_key: [{:awscli, "default", 1_800}],
  awscli_auth_adapter: ExAws.STS.AuthCache.AssumeRoleCredentialsAdapter,
  awscli_credentials: %{
    "default" => %{
      role_arn: {:system, "AWS_ROLE_ARN"},
      access_key_id: {:system, "AWS_ACCESS_KEY_ID"},
      secret_access_key: {:system, "AWS_SECRET_ACCESS_KEY"},
      source_profile: "default"
    }
  }
```

### Using Web Identity tokens from ENV vars

Similarly, it is possible to use a web identity token to perform the assume role operation. It currently uses the following env vars to obtain it:

`AWS_WEB_IDENTITY_TOKEN_FILE`: path of the file with the web identity token
`AWS_ROLE_ARN`: role to be assumed

```elixir
config :ex_aws,
  secret_access_key: [{:awscli, "profile_name", 1_800}],
  access_key_id: [{:awscli, "profile_name", 1_800}],
  awscli_auth_adapter: ExAws.STS.AuthCache.AssumeRoleWebIdentityAdapter
```

## License

The MIT License (MIT)

Copyright (c) 2014 CargoSense, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v2.3.0 - 2020-05-15

- Removed hardcoded `Poison` dependency, now uses configured `json_codec` within the parent `ex_aws` module.

## v2.2.0 - 2020-05-14

- Added ability to allow STS credentials to be injected by configuration

## v2.1.0 - 2020-04-06

* Changed implementation to use `File.read!` for user-friendly error stack trace
* Fixed callback signature warning
* Fixed issue caused by changing key order in map
* Removed unused entry point that doesn't implement callback behaviour
* Added adapter for using Web Identity tokens when assuming role
* Fixed bug in XML parsing paths
* Added missing typespec to `get_caller_identity/0`
* Fixed typespec on `assume_role_with_saml/4`
* Added `AssumeRoleWithWebIdentity`, `GetAccessKeyInfo`, and `AssumeRoleWithSAML`

## v2.0.1 - 2019-07-17

* Fixed issue with `AssumeRoleCredentialsAdapter`

## v2.0.0 - 2017-11-10

- Major Project Split. Please see the main ExAws repository for previous changelogs.


test/test_helper.exs
Mox.defmock(ExAws.Request.HttpMock, for: ExAws.Request.HttpClient)

ExUnit.start()


test/lib/sts_test.exs
defmodule ExAws.STSTest do
  use ExUnit.Case, async: true

  alias ExAws.STS

  test "#get_caller_identity" do
    version = "2011-06-15"

    expected = %{
      "Action" => "GetCallerIdentity",
      "Version" => version
    }

    assert expected == STS.get_caller_identity().params
  end

  test "#assume_role" do
    version = "2011-06-15"
    arn = "1111111/test_role"
    name = "test role"
    external_id = "test id"

    expected = %{
      "Action" => "AssumeRole",
      "RoleSessionName" => name,
      "RoleArn" => arn,
      "Version" => version,
      "ExternalId" => external_id
    }

    assert expected == STS.assume_role(arn, name, external_id: external_id).params
  end

  test "#assume_role_with_web_identity" do
    version = "2011-06-15"
    arn = "1111111/test_role"
    name = "test role"
    token = "atoken"

    expected = %{
      "Action" => "AssumeRoleWithWebIdentity",
      "RoleSessionName" => name,
      "RoleArn" => arn,
      "WebIdentityToken" => token,
      "Version" => version
    }

    assert expected == STS.assume_role_with_web_identity(arn, name, token).params
  end

  test "#assume_role_with_saml" do
    version = "2011-06-15"
    principal_arn = "1111111/test_principal"
    role_arn = "1111111/test_role"
    saml_assertion = "assertioncontent" |> Base.encode64()

    expected = %{
      "Action" => "AssumeRoleWithSAML",
      "PrincipalArn" => principal_arn,
      "RoleArn" => role_arn,
      "SAMLAssertion" => saml_assertion,
      "Version" => version
    }

    assert expected == STS.assume_role_with_saml(principal_arn, role_arn, saml_assertion).params
  end

  test "#decode_authorization_message" do
    version = "2011-06-15"
    message = "msgcontent"

    expected = %{
      "Action" => "DecodeAuthorizationMessage",
      "EncodedMessage" => message,
      "Version" => version
    }

    assert expected == STS.decode_authorization_message(message).params
  end

  test "#get_access_key_info" do
    version = "2011-06-15"
    key_id = "AKIAI44QH8DHBEXAMPLE"

    expected = %{
      "Action" => "GetAccessKeyInfo",
      "AccessKeyId" => key_id,
      "Version" => version
    }

    assert expected == STS.get_access_key_info(key_id).params
  end

  test "#get_federation_token" do
    version = "2011-06-15"
    duration = 900
    name = "Bob"

    policy = %{
      "Statement" => [
        %{
          "Sid" => "Stmt1",
          "Effect" => "Allow",
          "Action" => "s3:*",
          "Resource" => "*"
        }
      ]
    }

    expected = %{
      "Action" => "GetFederationToken",
      "DurationSeconds" => duration,
      "Name" => name,
      "Policy" => Jason.encode!(policy),
      "Version" => version
    }

    opts = [duration: duration, policy: policy]

    assert expected == STS.get_federation_token(name, opts).params
  end

  test "#get_session_token" do
    version = "2011-06-15"
    duration = 900

    expected = %{
      "Action" => "GetSessionToken",
      "DurationSeconds" => duration,
      "Version" => version
    }

    assert expected == STS.get_session_token(duration: duration).params
  end
end


test/lib/parsers_test.exs
defmodule ExAws.STS.ParsersTest do
  use ExUnit.Case, async: true

  alias ExAws.STS.Parsers

  def parse_mock_response(action, arity) when is_atom(action) do
    p = Path.join([__DIR__, "fixtures", "#{to_string(action)}_response.xml"])
    mock_response = {:ok, %{body: File.read!(p)}}

    case arity do
      2 ->
        Parsers.parse(mock_response, action)

      3 ->
        config = [json_codec: Jason]

        Parsers.parse(mock_response, action, config)
    end
  end

  @actions [
    assume_role: 2,
    assume_role_with_s_a_m_l: 2,
    assume_role_with_web_identity: 2,
    decode_authorization_message: 3,
    get_access_key_info: 2,
    get_caller_identity: 2,
    get_session_token: 2,
    get_federation_token: 2
  ]

  # Build tests for all actions and verifies that mock responses were parsed
  for {action, arity} <- @actions do
    @tag action: action
    test "parses `:#{action}` using mock response" do
      assert {:ok, %{body: body}} = parse_mock_response(unquote(action), unquote(arity))

      for {_, value} <- body do
        assert value && value != "", "did not parse value for response"
      end
    end
  end
end


mix.exs
defmodule ExAws.STS.Mixfile do
  use Mix.Project

  @version "2.3.0"
  @service "sts"
  @url "https://github.com/ex-aws/ex_aws_#{@service}"
  @name __MODULE__ |> Module.split() |> Enum.take(2) |> Enum.join(".")

  def project do
    [
      app: :ex_aws_sts,
      version: @version,
      elixir: "~> 1.5",
      elixirc_paths: elixirc_paths(Mix.env()),
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      name: @name,
      package: package(),
      docs: docs()
    ]
  end

  defp package do
    [
      description: "#{@name} service package",
      files: ["lib", "config", "mix.exs", "README*", "CHANGELOG*", "CONTRIBUTING*"],
      maintainers: ["Ben Wilson"],
      licenses: ["MIT"],
      links: %{
        Changelog: "https://hexdocs.pm/ex_aws_sts/changelog.html",
        GitHub: @url
      }
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:mox, ">= 0.0.3", only: :test},
      {:briefly, ">= 0.0.3", only: :test},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:hackney, ">= 0.0.0", only: [:dev, :test]},
      {:sweet_xml, ">= 0.0.0", only: [:dev, :test]},
      {:jason, ">= 0.0.0", only: [:dev, :test]},
      {:xml_builder, "~> 2.1", only: [:test]},
      ex_aws()
    ]
  end

  defp docs do
    [
      extras: ["CHANGELOG.md", "CONTRIBUTING.md", "README.md"],
      main: "readme",
      source_ref: "v#{@version}",
      source_url: @url,
      formatters: ["html"]
    ]
  end

  defp ex_aws() do
    case System.get_env("AWS") do
      "LOCAL" -> {:ex_aws, path: "../ex_aws"}
      _ -> {:ex_aws, "~> 2.2"}
    end
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :ex_aws_sts, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:ex_aws_sts, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/ex_aws/sts/parsers.ex
if Code.ensure_loaded?(SweetXml) do
  defmodule ExAws.STS.Parsers do
    import SweetXml, only: [sigil_x: 2]

    def parse({:ok, %{body: xml} = resp}, :assume_role) do
      parsed_body =
        xml
        |> SweetXml.xpath(~x"//AssumeRoleResponse",
          access_key_id: ~x"./AssumeRoleResult/Credentials/AccessKeyId/text()"s,
          secret_access_key: ~x"./AssumeRoleResult/Credentials/SecretAccessKey/text()"s,
          session_token: ~x"./AssumeRoleResult/Credentials/SessionToken/text()"s,
          expiration: ~x"./AssumeRoleResult/Credentials/Expiration/text()"s,
          assumed_role_id: ~x"./AssumeRoleResult/AssumedRoleUser/AssumedRoleId/text()"s,
          assumed_role_arn: ~x"./AssumeRoleResult/AssumedRoleUser/Arn/text()"s,
          request_id: request_id_xpath()
        )

      {:ok, Map.put(resp, :body, parsed_body)}
    end

    def parse({:ok, %{body: xml} = resp}, :assume_role_with_web_identity) do
      parsed_body =
        xml
        |> SweetXml.xpath(~x"//AssumeRoleWithWebIdentityResponse",
          access_key_id: ~x"./AssumeRoleWithWebIdentityResult/Credentials/AccessKeyId/text()"s,
          secret_access_key:
            ~x"./AssumeRoleWithWebIdentityResult/Credentials/SecretAccessKey/text()"s,
          session_token: ~x"./AssumeRoleWithWebIdentityResult/Credentials/SessionToken/text()"s,
          expiration: ~x"./AssumeRoleWithWebIdentityResult/Credentials/Expiration/text()"s,
          assumed_role_id:
            ~x"./AssumeRoleWithWebIdentityResult/AssumedRoleUser/AssumedRoleId/text()"s,
          assumed_role_arn: ~x"./AssumeRoleWithWebIdentityResult/AssumedRoleUser/Arn/text()"s,
          request_id: request_id_xpath()
        )
      {:ok, Map.put(resp, :body, parsed_body)}
    end

    def parse({:ok, %{body: xml} = resp}, :get_access_key_info) do
      parsed_body =
        SweetXml.xpath(xml, ~x"//GetAccessKeyInfoResponse",
          account: ~x"./GetAccessKeyInfoResult/Account/text()"s,
          request_id: request_id_xpath()
        )

      {:ok, Map.put(resp, :body, parsed_body)}
    end

    def parse({:ok, %{body: xml} = resp}, :assume_role_with_s_a_m_l) do
      parsed_body =
        xml
        |> SweetXml.xpath(~x"//AssumeRoleWithSAMLResponse",
          access_key_id: ~x"./AssumeRoleResult/Credentials/AccessKeyId/text()"s,
          secret_access_key: ~x"./AssumeRoleResult/Credentials/SecretAccessKey/text()"s,
          session_token: ~x"./AssumeRoleResult/Credentials/SessionToken/text()"s,
          expiration: ~x"./AssumeRoleResult/Credentials/Expiration/text()"s,
          assumed_role_id: ~x"./AssumeRoleResult/AssumedRoleUser/AssumedRoleId/text()"s,
          assumed_role_arn: ~x"./AssumeRoleResult/AssumedRoleUser/Arn/text()"s,
          request_id: request_id_xpath()
        )

      {:ok, Map.put(resp, :body, parsed_body)}
    end

    def parse({:ok, %{body: xml} = resp}, :get_caller_identity) do
      parsed_body =
        SweetXml.xpath(xml, ~x"//GetCallerIdentityResponse",
          arn: ~x"./GetCallerIdentityResult/Arn/text()"s,
          user_id: ~x"./GetCallerIdentityResult/UserId/text()"s,
          account: ~x"./GetCallerIdentityResult/Account/text()"s,
          request_id: request_id_xpath()
        )

      {:ok, Map.put(resp, :body, parsed_body)}
    end

    def parse({:ok, %{body: xml} = resp}, :get_federation_token) do
      parsed_body =
        xml
        |> SweetXml.xpath(~x"//GetFederationTokenResponse",
          access_key_id: ~x"./GetFederationTokenResult/Credentials/AccessKeyId/text()"s,
          secret_access_key: ~x"./GetFederationTokenResult/Credentials/SecretAccessKey/text()"s,
          session_token: ~x"./GetFederationTokenResult/Credentials/SessionToken/text()"s,
          expiration: ~x"./GetFederationTokenResult/Credentials/Expiration/text()"s,
          request_id: request_id_xpath()
        )

      {:ok, Map.put(resp, :body, parsed_body)}
    end

    def parse({:ok, %{body: xml} = resp}, :get_session_token) do
      parsed_body =
        xml
        |> SweetXml.xpath(~x"//GetSessionTokenResponse",
          access_key_id: ~x"./GetSessionTokenResult/Credentials/AccessKeyId/text()"s,
          secret_access_key: ~x"./GetSessionTokenResult/Credentials/SecretAccessKey/text()"s,
          session_token: ~x"./GetSessionTokenResult/Credentials/SessionToken/text()"s,
          expiration: ~x"./GetSessionTokenResult/Credentials/Expiration/text()"s,
          request_id: request_id_xpath()
        )

      {:ok, Map.put(resp, :body, parsed_body)}
    end

    def parse({:error, {type, http_status_code, %{body: xml}}}, _) do
      parsed_body =
        xml
        |> SweetXml.xpath(~x"//ErrorResponse",
          request_id: ~x"./RequestId/text()"s,
          type: ~x"./Error/Type/text()"s,
          code: ~x"./Error/Code/text()"s,
          message: ~x"./Error/Message/text()"s,
          detail: ~x"./Error/Detail/text()"s
        )

      {:error, {type, http_status_code, parsed_body}}
    end

    def parse(val, _), do: val

    def parse({:ok, %{body: xml} = resp}, :decode_authorization_message, config) do
      parsed_body =
        xml
        |> SweetXml.xpath(~x"//DecodeAuthorizationMessageResponse",
          decoded_message: ~x"./DecodedMessage/text()"s,
          request_id: ~x"./RequestId/text()"s
        )
        |> Map.update!(:decoded_message, fn msg -> config[:json_codec].decode!(msg) end)

      {:ok, Map.put(resp, :body, parsed_body)}
    end

    defp request_id_xpath do
      ~x"./ResponseMetadata/RequestId/text()"s
    end
  end
else
  defmodule ExAws.STS.Parsers do
    def parse(val, _), do: val
    def parse(val, _, _), do: val
  end
end


lib/ex_aws/sts/auth_cache/assume_role_credentials_adapter.ex
defmodule ExAws.STS.AuthCache.AssumeRoleCredentialsAdapter do
  @moduledoc """
  Provides a custom Adapter that intercepts ExAWS configuration
  which uses Role ARN + Source Profile for authentication.
  """

  @behaviour ExAws.Config.AuthCache.AuthConfigAdapter

  @impl true
  def adapt_auth_config(auth, profile, expiration) do
    adapt_auth_config(auth, profile, expiration, &load_credentials/1)
  end

  def adapt_auth_config(%{source_profile: source_profile} = auth, _, expiration, loader) do
    source_profile_auth = loader.(source_profile)
    get_security_credentials(auth, source_profile_auth, expiration)
  end

  def adapt_auth_config(%{credential_source: "Ec2InstanceMetadata"} = auth, _, expiration, _loader) do
    instance_auth =
      ExAws.Config.new(:sts, access_key_id: "dummy", secret_access_key: "dummy")
      |>ExAws.InstanceMeta.security_credentials()

    get_security_credentials(auth, instance_auth, expiration)
  end

  def adapt_auth_config(auth, _, _, _), do: auth

  defp get_security_credentials(auth, source_profile_auth, expiration) do
    duration = credential_duration_seconds(expiration)
    role_session_name = Map.get(auth, :role_session_name, "default_session")

    assume_role_options =
      case auth do
        %{external_id: external_id} -> [duration: duration, external_id: external_id]
        _ -> [duration: duration]
      end

    assume_role_request =
      ExAws.STS.assume_role(auth.role_arn, role_session_name, assume_role_options)

    assume_role_config = ExAws.Config.new(:sts, source_profile_auth)

    with {:ok, result} <- ExAws.request(assume_role_request, assume_role_config) do
      %{
        access_key_id: result.body.access_key_id,
        secret_access_key: result.body.secret_access_key,
        security_token: result.body.session_token,
        role_arn: auth.role_arn,
        role_session_name: role_session_name
      }
    else
      {:error, reason} ->
        {:error, reason}
    end
  end

  defp credential_duration_seconds(expiration_ms) do
    # assume_role accepts a duration between 900 and 3600 seconds
    # We're adding a buffer to make sure the credentials live longer than
    # the refresh interval.
    {min, max, buffer} = {900, 3600, 5}
    seconds = div(expiration_ms, 1000) + buffer
    Enum.max([Enum.min([max, seconds]), min])
  end

  defp load_credentials(profile), do: ExAws.Config.awscli_auth_credentials(profile)
end


lib/ex_aws/sts/auth_cache/assume_role_credentials_provider.ex
defmodule ExAws.STS.AuthCache.AssumeRoleCredentialsProvider do
  @moduledoc false

  @behaviour ExAws.Config.AuthCache.AuthConfigProvider

  @impl true
  def auth_config_for(profile, _expiration) do
    auth = ExAws.CredentialsIni.security_credentials(profile)

    if %{source_profile: source_profile} = auth do
      auth = ExAws.CredentialsIni.security_credentials(source_profile)
    end

    security_credentials_for(auth, expiration)
  end

  defp security_credentials_for(%{role_arn: role_arn} = auth, expiration) do
    assume_role_options = [
      duration: credential_duration_seconds(expiration)
    ]

    if auth[:external_id] do
      assume_role_options = Keyword.put(assume_role_options, :external_id, auth[:external_id])
    end

    role_session_name = auth[:role_session_name] || "default_session"

    {:ok, result} = role_arn
                    |> ExAws.STS.assume_role(role_session_name, assume_role_options)
                    |> ExAws.Operation.perform(ExAws.Config.new(:sts))

    %{
      access_key_id: result.body.access_key_id,
      secret_access_key: result.body.secret_access_key,
      security_token: result.body.session_token,
      expiration: result.body.expiration
    }
  end

  defp security_credentials_for(auth, _expiration), do: auth

  defp credential_duration_seconds(expiration_ms) do
    # assume_role accepts a duration between 900 and 3600 seconds
    # We're adding a buffer to make sure the credentials live longer than
    # the refresh interval.
    {min, max, buffer} = {900, 3600, 5}
    seconds = div(expiration_ms, 1000) + buffer
    Enum.max([ Enum.min([max, seconds]), min ])
  end
end


lib/ex_aws/sts/auth_cache/assume_role_web_identity_adapter.ex
defmodule ExAws.STS.AuthCache.AssumeRoleWebIdentityAdapter do
  @moduledoc """
  Provides a custom Adapter that intercepts ExAWS configuration
  which uses Role ARN + Web Identity Tokens for authentication.
  """

  @behaviour ExAws.Config.AuthCache.AuthConfigAdapter

  @impl true
  def adapt_auth_config(config, profile, expiration) do
    adapt_auth_config(config, profile, expiration, &loader/1)
  end

  def adapt_auth_config(config, _profile, expiration, loader) do
    auth = Map.merge(config, loader.(config))
    get_security_credentials(auth, expiration || 30_000)
  end

  defp get_security_credentials(auth, expiration) do
    duration = credential_duration_seconds(expiration)

    assume_role_request =
      ExAws.STS.assume_role_with_web_identity(
        auth.role_arn,
        auth.role_session_name,
        auth.web_identity_token,
        duration: duration
      )

    config_overrides =
      case auth.use_regional_endpoints do
        false -> auth |> Map.put(:host, "sts.amazonaws.com")
        true -> auth
      end

    with {:ok, result} <- ExAws.request(assume_role_request, config_overrides) do
      %{
        access_key_id: result.body.access_key_id,
        secret_access_key: result.body.secret_access_key,
        security_token: result.body.session_token,
        role_arn: auth.role_arn,
        role_session_name: auth.role_session_name,
        expiration: expiration
      }
    else
      {:error, reason} ->
        {:error, reason}
    end
  end

  defp credential_duration_seconds(expiration_ms) do
    # assume_role accepts a duration between 900 and 3600 seconds
    # We're adding a buffer to make sure the credentials live longer than
    # the refresh interval.
    {min, max, buffer} = {900, 3600, 5}
    seconds = div(expiration_ms, 1000) + buffer
    Enum.max([Enum.min([max, seconds]), min])
  end

  defp loader(config) do
    %{
      role_arn: env_role_arn(config),
      role_session_name: role_session_name(config),
      web_identity_token: web_identity_token(config),
      region: aws_region(config),
      use_regional_endpoints: use_regional_endpoints(config),
      # necessary for now due to how ExAws.request() works
      access_key_id: "dummy",
      # necessary for now due to how ExAws.request() works
      secret_access_key: "dummy"
    }
  end

  defp web_identity_token(config) do
    config
    |> web_identity_token_file()
    |> File.read!()
  end

  defp web_identity_token_file(config) do
    config[:web_identity_token_file] || System.get_env("AWS_WEB_IDENTITY_TOKEN_FILE")
  end

  defp env_role_arn(config) do
    config[:role_arn] || System.get_env("AWS_ROLE_ARN")
  end

  defp role_session_name(config) do
    config[:role_session_name] || "default_session"
  end

  defp use_regional_endpoints(config) do
    regional_endpoints =
      config[:sts_regional_endpoints] || System.get_env("AWS_STS_REGIONAL_ENDPOINTS")

    case regional_endpoints do
      "regional" -> true
      _ -> false
    end
  end

  defp aws_region(config) do
    System.get_env("AWS_REGION") || config[:region]
  end
end


lib/ex_aws/sts.ex
defmodule ExAws.STS do
  @moduledoc """
  Operations on AWS STS.

  See https://docs.aws.amazon.com/STS/latest/APIReference/API_Operations.html
  """

  @type policy :: %{
          binary => :all
        }

  @type assume_role_opt ::
          {:duration, pos_integer}
          | {:serial_number, binary}
          | {:token_code, binary}
          | {:external_id, binary}
          | {:policy, policy}

  @doc """
  Assume Role.
  """
  @spec assume_role(role_arn :: String.t(), role_session_name :: String.t(), [assume_role_opt]) ::
          ExAws.Operation.Query.t()
  def assume_role(role_arn, role_session_name, opts \\ []) do
    params =
      parse_opts(opts)
      |> Map.put("RoleArn", role_arn)
      |> Map.put("RoleSessionName", role_session_name)

    request(:assume_role, params)
  end

  @type assume_role_with_web_identity_opt ::
          {:duration, pos_integer}
          | {:provider_id, binary}
          | {:policy, policy}

  @doc """
  Assume Role with Web Identity.
  """
  @spec assume_role_with_web_identity(
          role_arn :: String.t(),
          role_session_name :: String.t(),
          web_identity_token :: String.t(),
          [assume_role_with_web_identity_opt]
        ) :: ExAws.Operation.Query.t()
  def assume_role_with_web_identity(role_arn, role_session_name, web_identity_token, opts \\ []) do
    params =
      parse_opts(opts)
      |> Map.put("RoleArn", role_arn)
      |> Map.put("RoleSessionName", role_session_name)
      |> Map.put("WebIdentityToken", web_identity_token)

    request(:assume_role_with_web_identity, params)
  end

  @type assume_role_with_saml_opt ::
          {:duration, pos_integer}
          | {:policy, policy}

  @doc """
  Assume Role with SAML.
  """
  @spec assume_role_with_saml(
          principal_arn :: String.t(),
          role_arn :: String.t(),
          saml_assertion :: String.t(),
          [assume_role_with_saml_opt]
        ) :: ExAws.Operation.Query.t()
  def assume_role_with_saml(principal_arn, role_arn, saml_assertion, opts \\ []) do
    params =
      parse_opts(opts)
      |> Map.put("PrincipalArn", principal_arn)
      |> Map.put("RoleArn", role_arn)
      |> Map.put("SAMLAssertion", saml_assertion)

    request(:assume_role_with_s_a_m_l, params)
  end

  @doc """
  Decode Authorization Message.
  """
  @spec decode_authorization_message(message :: String.t()) :: ExAws.Operation.Query.t()
  def decode_authorization_message(message) do
    request(:decode_authorization_message, %{"EncodedMessage" => message}, %{
      parser: &ExAws.STS.Parsers.parse/3
    })
  end

  @doc """
  Get Access Key Info.
  """
  @spec get_access_key_info(key_id :: String.t()) :: ExAws.Operation.Query.t()
  def get_access_key_info(key_id) do
    request(:get_access_key_info, %{"AccessKeyId" => key_id})
  end

  @doc """
  Get Caller Identity.
  """
  @spec get_caller_identity() :: ExAws.Operation.Query.t()
  def get_caller_identity() do
    request(:get_caller_identity, %{})
  end

  @type get_federation_token_opt :: {:duration, pos_integer} | {:policy, policy}

  @doc """
  Get Federation Token.
  """
  @spec get_federation_token(name :: String.t(), [get_federation_token_opt]) ::
          ExAws.Operation.Query.t()
  def get_federation_token(name, opts \\ []) do
    params =
      parse_opts(opts)
      |> Map.put("Name", name)

    request(:get_federation_token, params)
  end

  @type get_session_token_opt ::
          {:duration, pos_integer}
          | {:serial_number, binary}
          | {:token_code, binary}

  @doc """
  Get Session Token.
  """
  @spec get_session_token([get_session_token_opt]) :: ExAws.Operation.Query.t()
  def get_session_token(opts \\ []) do
    params = parse_opts(opts)

    request(:get_session_token, params)
  end

  ## Request
  ######################

  defp request(action, params, overrides \\ %{}) do
    action_string = action |> Atom.to_string() |> Macro.camelize()

    params =
      Map.merge(params, %{
        "Version" => "2011-06-15",
        "Action" => action_string
      })

    %ExAws.Operation.Query{
      path: "/",
      params: params,
      service: :sts,
      action: action,
      parser: &ExAws.STS.Parsers.parse/2
    }
    |> struct(overrides)
  end

  defp parse_opts(opts) do
    Enum.reduce(opts, %{}, fn item, acc -> parse_opt(acc, item) end)
  end

  defp parse_opt(opts, {:duration, val}), do: Map.put(opts, "DurationSeconds", val)
  defp parse_opt(opts, {:token_code, val}), do: Map.put(opts, "TokenCode", val)
  defp parse_opt(opts, {:serial_number, val}), do: Map.put(opts, "SerialNumber", val)
  defp parse_opt(opts, {:provider_id, val}), do: Map.put(opts, "ProviderId", val)
  defp parse_opt(opts, {:external_id, val}), do: Map.put(opts, "ExternalId", val)
  defp parse_opt(opts, {:policy, val}), do: Map.put(opts, "Policy", json_codec().encode!(val))

  defp json_codec(), do: ExAws.Config.build_base(:sts) |> Map.get(:json_codec)
end
